Dockerfile

```diff
@@ -1,35 +1,51 @@
+# ---- Base Node ----
 FROM node:18.13.0-bullseye-slim as base
 
-RUN set -ex; \
-    apt update -y; \
-    apt install ca-certificates -y; \
-    update-ca-certificates
-
-FROM base as build
-
-WORKDIR /app
-
-RUN apt install make gcc g++ python3 git openssl -y;
+RUN mkdir -p /home/node/app
+RUN chown -R node:node /home/node && chmod -R 770 /home/node
+WORKDIR /home/node/app
+
+# ---- Dependencies ----
+FROM base AS dependencies
+
+COPY --chown=node:node ./package.json ./package.json prisma ./
+COPY --chown=node:node ./package-lock.json ./package-lock.json
+RUN apt-get update -y && \
+    apt-get install -y --no-install-recommends \
+      ca-certificates \
+      dumb-init \
+      make \
+      gcc \
+      g++ \
+      python3 \
+      git \
+      openssl
+USER node
+# install production dependencies
+RUN npm ci
+# separate production node_modules
+RUN cp -R node_modules prod_node_modules
+# install ALL node_modules, including 'devDependencies'
+RUN npm install
 
-COPY --chown=node:node package.json package-lock.json prisma ./
-RUN set -ex; \
-    npm install --ignore-scripts; \
-    npx prisma generate
+# ---- Build ----
+FROM dependencies as build
 
 COPY --chown=node:node . .
-RUN set -ex; \
-    npm run build; \
-    NODE_ENV=production npm prune
-
-FROM base as app
+RUN npx prisma generate && \
+    npm run build
 
-WORKDIR /app
+# ---- Release ----
+FROM base as release
 
-COPY --from=build /app/dist ./
-COPY --from=build /app/prisma ./prisma
-COPY --from=build /app/node_modules ./node_modules
+COPY --from=build /usr/bin/dumb-init /usr/bin/dumb-init
+COPY --from=build /home/node/app/dist ./
+COPY --from=build /home/node/app/prisma ./prisma
+COPY --from=build /home/node/app/prod_node_modules ./node_modules
+COPY --from=build /home/node/app/src ./src
 
+ENV NODE_ENV=production
 ENV NODE_PATH=.
 
 USER node
-CMD [ "node", "/app/src/server.js" ]
+CMD [ "dumb-init", "node", "/home/node/app/src/server.js" ]
```

Comment:
What's the rationale behind using dumb-init?

Reply:
> Node.js was not designed to run as PID 1 which leads to unexpected behaviour when running inside of Docker. For example, a Node.js process running as PID 1 will not respond to SIGINT (CTRL-C) and similar signals. As of Docker 1.13, you can use the --init flag to wrap your Node.js process with a [lightweight init system](https://github.com/krallin/tini) that properly handles running as PID 1.

https://github.com/nodejs/docker-node/blob/main/docs/BestPractices.md#handling-kernel-signals

Reply:
The official image doesn't use dumb-init though https://github.com/nodejs/docker-node/blob/16b241361a3bef1f40ec3ef96afa642cab7a336e/18/bullseye-slim/Dockerfile#L92

Reply:
I think it's better not to rely on Docker container runtime built-ins, as those are [not available in Kubernetes for example](https://kubernetes.io/blog/2022/02/17/dockershim-faq/).